### PR TITLE
Proposed fix for #3624 NPE at RangeQueryRewriter.rewriteLocationStep

### PR DESCRIPTION
**Pull request by dizzzz**
Proposed fix for #3624 NPE at org.exist.xquery.modules.range.RangeQueryRewriter.rewriteLocationStep
Handle all operators

Help needed for tests

### CODE REVIEW

**Comment by joewiz** [comment]
I can confirm that this fixes the issue I reported in #3624. Thank you, @dizzzz!

**Comment by adamretter** [comment]
@dizzzz Interesting stuff... I have some questions ;-)

1. Previously this was limited to `fn:matches` but now it is only restricted to a Function that implements `Lookup`. Joe mentioned he did not see the problem in 5.2.0 - but this code would not have changed between 5.3.0 and 5.2.0.

2. What was the expression that was causing the issue?

3. This code can still cause an NPE:

```
Lookup func = Lookup.create(function.getContext(), operator, path);
func.setArguments(eqArgs);
```

The `Lookup.Create` function can return `null`. So the value of `func` should probably be checked/logged before it is used or not.

**Comment by dizzzz** [comment]
`eq` came in as option instead of 'matches', then I realized that only one item was handled. In other cases than `matches` NULL was returned, yielding in  NPE; 

Instead of preventing the `eq` NPE, I thought it would make sense just to handle all `enum` options 👍 That should be OK as all options should accept 2 parameters,

Hmmm did I miss a potential NPE? I thought I covered it all....

**Comment by adamretter** [comment]
Hmmm, shouldn't `eq` be handled just above in the `if (expression instanceof GeneralComparison) {` branch?

**Comment by dizzzz** [comment]
Hmmmm  you are right on this I guess. good spot. hmmmm

**Comment by dizzzz** [comment]
What shall I do? close PR?

File: extensions/indexes/range/src/main/java/org/exist/xquery/modules/range/RangeQueryRewriter.java [hunk 1]
```diff
@@ -138,13 +138,17 @@ protected static Lookup rewrite(Expression expression, NodePath path) throws XPa
...
             if (function instanceof Lookup) {
-                if (function.isCalledAs("matches")) {
-                    eqArgs.add(function.getArgument(0));
-                    eqArgs.add(function.getArgument(1));
-                    Lookup func = Lookup.create(function.getContext(), RangeIndex.Operator.MATCH, path);
-                    func.setArguments(eqArgs);
-                    return func;
+                final RangeIndex.Operator operator = RangeIndex.Operator.getByName(function.getName().getLocalPart());
+
+                if(function==null){
```
Comment (wolfgangmm) [inline extensions/indexes/range/src/main/java/org/exist/xquery/modules/range/RangeQueryRewriter.java:143]: The null check is too late here. Should go further up.

File: extensions/indexes/range/src/main/java/org/exist/xquery/modules/range/RangeQueryRewriter.java [hunk 2]
```diff
@@ -138,13 +138,17 @@ protected static Lookup rewrite(Expression expression, NodePath path) throws XPa
...
+                    return null;
                 }
+                
+                eqArgs.add(function.getArgument(0));
+                eqArgs.add(function.getArgument(1));
+                Lookup func = Lookup.create(function.getContext(), operator, path);
+                func.setArguments(eqArgs);
```
Comment (wolfgangmm) [inline extensions/indexes/range/src/main/java/org/exist/xquery/modules/range/RangeQueryRewriter.java:150]: Unlikely, but func could be null. Check here?

**Review by wolfgangmm — CHANGES_REQUESTED** [review]
I debugged the original issue myself and came to the conclusion that Dannes' proposed fix is basically correct. What we have here is an expression which has already been partially rewritten (the `GeneralComparison` was replaced by `range:eq`). I'm not sure where this partial optimization was applied - probably an earlier traversal of the expression (we would need a much more reduced test case to know), but it is certainly a possible case which should not lead to an exception.

As Dannes pointed out, `RangeQueryRewriter.rewrite` only handled `matches` (stupid me!). It should handle all functions implementing `Lookup`.

File: extensions/indexes/range/src/main/java/org/exist/xquery/modules/range/RangeQueryRewriter.java (hunk 1 above)
Comment (dizzzz) [reply extensions/indexes/range/src/main/java/org/exist/xquery/modules/range/RangeQueryRewriter.java]: done

File: extensions/indexes/range/src/main/java/org/exist/xquery/modules/range/RangeQueryRewriter.java (hunk 2 above)
Comment (dizzzz) [reply extensions/indexes/range/src/main/java/org/exist/xquery/modules/range/RangeQueryRewriter.java:148]: done

**Comment by dizzzz** [comment]
@wolfgangmm @adamretter ready to go

**Review by joewiz — APPROVED** [review]
@dizzzz I can confirm that the updated PR still fixes the issue I reported in #3624. Thank you!

**Review by wolfgangmm — APPROVED** [review]

**Comment by joewiz** [comment]
@dizzzz Thank you so much for this fix! And thanks to @adamretter and @wolfgangmm for your reviews.

**Comment by adamretter** [comment]
I just dug into this further to try and understand why this fix works. I reduced Dannes's changes to the absolute minimum to pass Joe's reported issue #3624, and was left with just this patch:

```
diff --git a/extensions/indexes/range/src/main/java/org/exist/xquery/modules/range/RangeQueryRewriter.java b/extensions/indexes/range/src/main/java/org/exist/xquery/modules/range/RangeQueryRewriter.java
index 04f3db1f89..f7d47facce 100644
--- a/extensions/indexes/range/src/main/java/org/exist/xquery/modules/range/RangeQueryRewriter.java
+++ b/extensions/indexes/range/src/main/java/org/exist/xquery/modules/range/RangeQueryRewriter.java
@@ -139,13 +139,12 @@ public class RangeQueryRewriter extends QueryRewriter {
             InternalFunctionCall fcall = (InternalFunctionCall) expression;
             Function function = fcall.getFunction();
             if (function instanceof Lookup) {
-                if (function.isCalledAs("matches")) {
                     eqArgs.add(function.getArgument(0));
                     eqArgs.add(function.getArgument(1));
-                    Lookup func = Lookup.create(function.getContext(), RangeIndex.Operator.MATCH, path);
+                    final RangeIndex.Operator operator = RangeIndex.Operator.getByName(function.getName().getLocalPart());
+                    Lookup func = Lookup.create(function.getContext(), operator, path);
                     func.setArguments(eqArgs);
                     return func;
-                }
             }
         }
         return null;
```

As @wolfgangmm hinted at previously, the statement `if (function.isCalledAs("matches")) {` meant that `RangeQueryRewriter#rewrite(Expression, NodePath)` would return `null`, if it was presented with an instance of the `Lookup` class (which is an expression that is optimized for the Range Index) that was not `range:matches#2`.

The more I think about this, the less sense the current code makes. If we already have an instance of `Lookup` then it is already optimized and so it doesn't need to be passed to `RangeQueryRewriter#rewrite` to be rewritten as a Lookup function, because it is already a Lookup function!

With that in mind I still think it likely that the current fix is addressing a symptom rather than the root cause of the problem, and that eXist-db is doing more work than necessary by replacing one optimized Lookup expression with exactly the same optimized Lookup expression. I will keep digging...

**Comment by adamretter** [comment]
> If we already have an instance of Lookup then it is already optimized and so it doesn't need to be passed to RangeQueryRewriter#rewrite to be rewritten as a Lookup function, because it is already a Lookup function!

We (Evolved Binary) have provided an updated PR that takes a more efficient path in - https://github.com/eXist-db/exist/pull/4882

**Pull request merged**